### PR TITLE
[Example] GatedGCN

### PR DESCRIPTION
**Pull request by paoxiaode**
## Description

Pytorch implementation of GatedGCN in paper [Residual Gated Graph ConvNets](https://arxiv.org/abs/1711.07553)

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this

### CODE REVIEW

File: examples/pytorch/GatedGCN/layer.py
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/layer.py:1]: What about moving this to https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv to make it a core DGL nn layer?

File: examples/pytorch/GatedGCN/layer.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+"""
+    GatedGCN: Residual Gated Graph ConvNets
+    An Experimental Study of Neural Networks for Variable Graphs (Xavier Bresson and Thomas Laurent)
```
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/layer.py:3]: max 80 chars.

File: examples/pytorch/GatedGCN/main.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    y_true = torch.cat(y_true, dim=0).numpy()
+    y_pred = torch.cat(y_pred, dim=0).numpy()
+
+    return evaluator.eval({"y_true": y_true, "y_pred": y_pred})["rocauc"]
+
+
+def main():
+    device = "cuda:0" if torch.cuda.is_available() else "cpu"
+
+    # load ogb dataset & evaluator
```
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/main.py:50]: For sentences:

\# Load ogb dataset & evaluator.

File: examples/pytorch/GatedGCN/main.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    )
+
+    test_loader = DataLoader(
+        dataset[split_idx["test"]],
+        batch_size=args.batch_size,
+        shuffle=False,
+        collate_fn=collate_dgl,
+    )
+
+    # load model
```
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/main.py:77]: ditto.

File: examples/pytorch/GatedGCN/main.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        t1 = time.time()
+        train_loss = train(model, device, train_loader, opt, loss_fn)
+        elapsed_time = time.time() - t1
+
+        train_auc = evaluate(model, device, train_loader, evaluator)
+        valid_auc = evaluate(model, device, valid_loader, evaluator)
+        test_auc = evaluate(model, device, test_loader, evaluator)
+
+        print(
+            f"Epoch {i} | Elapsed Time {elapsed_time:.2f} | Train Loss: {train_loss:.4f} | Train Auc: {train_auc:.4f} | Valid Auc: {valid_auc:.4f} | Test Auc: {test_auc:.4f}"
```
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/main.py:101]: Follow the pattern here, so it is easier to write regression test for this example:
https://github.com/dmlc/dgl/blob/f41934df87c5c38ee0286ce6807fb8eedcf78ac1/examples/sparse/sign.py#L89

File: examples/pytorch/GatedGCN/main.py
Comment (frozenbugs) [inline examples/pytorch/GatedGCN/main.py:1]: Combine the main.py and model.py and move it to examples/core/gated_gcn.py

Or if you want to separate them, then move them to examples/core/gated_gcn/...

File: examples/pytorch/GatedGCN/layer.py
Comment (paoxiaode) [reply examples/pytorch/GatedGCN/layer.py:1]: I think so, this should be in conv folder, I will move it.

**Comment by paoxiaode** [comment]
@frozenbugs

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
LGTM, @mufeili please double check the logic.

**Comment by mufeili** [comment]
Could you move the logic of `GatedGCNConv` to another PR? It might be easier to merge that PR first.

This PR also lacks the unit tests related to this module. Please add a file `dgl/tests/python/pytorch/nn/conv/test_gatedgcnconv.py`. For an example of unit tests, see https://github.com/dmlc/dgl/pull/3934.

You also need to add GatedGCNConv in https://github.com/dmlc/dgl/blob/master/docs/source/api/python/nn-pytorch.rst

**Comment by paoxiaode** [comment]
need to merge https://github.com/dmlc/dgl/pull/5654 first

**Pull request closed without merging**